### PR TITLE
Reduce max_cache_pos to reduce Baichuan2-13B memory

### PR DESCRIPTION
**Pull request by hkvision**

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 1]
```diff
@@ -370,6 +370,14 @@ def convert_forward(m, target_m, new_forward):
         convert_forward(sub_m, target_m, new_forward)
 
 
+def replace_func(m, target_m, func_name, new_func):
+    for _, sub_m in m.named_children():
+        if isinstance(sub_m, target_m):
+            bound_method = new_func.__get__(sub_m, sub_m.__class__)
+            setattr(sub_m, func_name, bound_method)
+        convert_forward(sub_m, target_m, new_func)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:378]: replace `replace_func`?

Comment (hkvision) [reply python/llm/src/bigdl/llm/transformers/convert.py:378]: oh, sorry, this is a typo

File: python/llm/src/bigdl/llm/transformers/models/baichuan2.py [hunk 2]
```diff
@@ -301,3 +301,94 @@ def baichuan_attention_forward_13b(
...
+def _buffered_future_mask(tensor, maxpos, alibi, attn_heads):
+    _future_mask = torch.triu(_fill_with_neg_inf(torch.zeros([maxpos, maxpos])), 1)
+    _future_mask = _future_mask.unsqueeze(0) + alibi
+    new_future_mask = _future_mask.to(tensor)
+    return new_future_mask[: tensor.shape[0] * attn_heads, :maxpos, :maxpos]
+
+
+def baichuan_13b_gen_alibi_mask(tensor, n_head, max_pos):
+    # May use fp16 for alibi mask to further reduce memory
+    slopes = torch.Tensor(_get_interleave(n_head))#.half()
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/baichuan2.py:336]: For now first disable fp16 mask since it will impact a bit of the performance.
The other changes won't hurt the performance.
@jason-dai Is it OK to merge this PR first after I test on MTL?
Will investigate if memory could be further reduced in a future PR.

**Review by lalalapotter — APPROVED** [review]
LGTM

File: python/llm/src/bigdl/llm/transformers/models/baichuan2.py (hunk 2 above)
Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/models/baichuan2.py]: Does 1K input work?

Comment (hkvision) [reply python/llm/src/bigdl/llm/transformers/models/baichuan2.py]: Previously worked, but it doesn't work starting from yesterday. After emptying mask cache, it can work now, see more details here: https://github.com/analytics-zoo/nano/issues/815#issuecomment-1866193709

File: python/llm/src/bigdl/llm/transformers/models/baichuan2.py [hunk 3]
```diff
@@ -301,3 +298,247 @@ def baichuan_attention_forward_13b(
...
+        inverted_mask = 1.0 - expanded_mask
+        inverted_mask = inverted_mask.masked_fill(
+            inverted_mask.to(torch.bool), torch.finfo(alibi_mask.dtype).min
+        )
+        attention_mask = inverted_mask + alibi_mask.unsqueeze(0)
+        if input_ids.device.type == "xpu":
+            del alibi_mask
+            del inverted_mask
+            del expanded_mask
+            torch.xpu.empty_cache()
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/baichuan2.py:468]: These mask variables are no longer used and deleting them can save some memory. Does it make sense? @jason-dai

**Comment by hkvision** [comment]
Since empty cache will hurt performance, remove this for now and merge this PR first.
